harness/determined/pytorch/deepspeed/_deepspeed_trial.py

```diff
@@ -0,0 +1,1054 @@
+        )
+        if self.is_chief:
+            all_num_validation_batches, all_validation_batch_size = zip(*all_tuples)
+            all_num_validation_batches = [le for le in all_num_validation_batches if le is not None]
+            if min(all_num_validation_batches) < max(all_num_validation_batches):
+                logging.warning(
+                    "Validation dataloader length inconsistent across ranks. "
+                    "Using the minimum for validation length."
+                )
+            self.num_validation_batches = min(all_num_validation_batches)
+            all_validation_batch_size = [le for le in all_validation_batch_size if le is not None]
+            if min(all_validation_batch_size) < max(all_validation_batch_size):
+                logging.warning(
+                    "Validation batch size inconsistent across ranks. "
+                    "Num inputs tracking for validation will be incorrect."
+                )
+            self.validation_batch_size = min(all_validation_batch_size)
+
+        (
+            self.num_validation_batches,
+            self.validation_batch_size,
+        ) = self.context.distributed._zmq_broadcast(
+            (self.num_validation_batches, self.validation_batch_size)
+        )
+
+    def run(self) -> None:
+        # We create the dataloading iterators here rather than in __init__ because we have to be
+        # careful to trigger its shutdown explicitly, to avoid hangs when the user is using
+        # multiprocessing-based parallelism for their dataloader.
+        #
+        # We create it before loading state because we don't want the training_iterator shuffling
+        # values after we load state.
+        self.training_iterator = (
+            iter(self.training_loader) if self.training_loader is not None else None
+        )
+        try:
+            # If a load path is provided load weights and restore the data location.
+            if self.env.latest_checkpoint is not None:
+                logging.info(f"Restoring trial from checkpoint {self.env.latest_checkpoint}")
+                with self.context._core.checkpointing.restore_path(
+                    self.env.latest_checkpoint
+                ) as load_path:
+                    self._load(pathlib.Path(load_path))
+
+            with self.prof:
+                for callback in self.callbacks.values():
+                    with self.prof.record_timing(
+                        f"callbacks.{callback.__class__.__name__}.on_training_start"
+                    ):
+                        callback.on_training_start()
+                self._run()
+
+        finally:
+            # Explicitly trigger the dataloader iterator shutdowns (which happens in __del__).
+            # See the rather long note in pytorch/torch/utils/data/dataloader.py.
+            if self.training_iterator is not None:
+                del self.training_iterator
+
+    def _run(self) -> None:
+        assert self.workloads is not None
```

Comment:
I would like to defer substituting the workload sequencer till we factor it out completely and replace it with something better for all trial interfaces

Reply:
> till we factor it out completely and replace it with something better for all trial interfaces

That's not going to happen.  TFKerasTrial is going to be replaced whole-hog with the real Keras training loop, and TFKerasTrial will continue to work with legacy code, and it will continue to use the WorkloadSequencer.  At the same time, we'll be writing a Trainer API for PyTorchTrial, and at that time we will hopefully yank the workload sequencer out of PyTorchTrial.

One thing that I think is extra unfortunate about the WorkloadSequencer is how unaware of dataloader epochs it is.  With DeepSpeed, is it easy to do true epoch-based training?  If we could do that, that would be a big improvement in terms of ease-of-explanability to our users I think.  Big enough to justify not using the WorkloadSequencer, IMO.

Reply:
Liam and I spoke offline; deepspeed does not really do "true epoch-based training" right now anyway.